File: src/whatsapp/meta.js
import crypto from 'node:crypto';
import { config } from '../config.js';
import { logger, redact } from '../utils/logger.js';

const graphUrl = (path) => `https://graph.facebook.com/${config.meta.graphVersion}/${path}`;

/**
 * Meta signs every webhook body with the app secret. Comparing with
 * timingSafeEqual avoids leaking the expected digest through timing.
 */
export function verifySignature(rawBody, signatureHeader, appSecret = config.meta.appSecret) {
  if (!appSecret) return false;
  const header = String(signatureHeader ?? '');
  if (!header.startsWith('sha256=')) return false;
  const expected = crypto.createHmac('sha256', appSecret).update(rawBody).digest('hex');
  const received = header.slice('sha256='.length);
  const expectedBuf = Buffer.from(expected, 'utf8');
  const receivedBuf = Buffer.from(received, 'utf8');
  if (expectedBuf.length !== receivedBuf.length) return false;
  return crypto.timingSafeEqual(expectedBuf, receivedBuf);
}

/** Flattens the Cloud API webhook envelope into a list of inbound messages. */
export function parseWebhook(body) {
  const out = [];
  for (const entry of body?.entry ?? []) {
    for (const change of entry?.changes ?? []) {
      const value = change?.value ?? {};
      const contacts = value.contacts ?? [];
      for (const message of value.messages ?? []) {
        const contact = contacts.find((item) => item?.wa_id === message.from) ?? contacts[0];
        const base = {
          id: message.id,
          from: message.from,
          name: contact?.profile?.name ?? '',
          timestamp: message.timestamp ? Number(message.timestamp) * 1000 : Date.now(),
          type: message.type,
          text: '',
          media: null,
        };

        switch (message.type) {
          case 'text':
            base.text = message.text?.body ?? '';
            break;
          case 'image':
            base.text = message.image?.caption ?? '';
            base.media = { id: message.image?.id, mediaType: message.image?.mime_type ?? 'image/jpeg' };
            break;
          case 'button':
            base.text = message.button?.text ?? '';
            break;
          case 'interactive':
            base.text =
              message.interactive?.button_reply?.title ?? message.interactive?.list_reply?.title ?? '';
            break;
          default:
            base.text = '';
        }
        out.push(base);
      }
    }
  }
  return out;
}

async function graphRequest(path, { method = 'POST', body, headers = {} } = {}) {
  const response = await fetch(graphUrl(path), {
    method,
    headers: {
      authorization: `Bearer ${config.meta.token}`,
      'content-type': 'application/json',
      ...headers,
    },
    body: body ? JSON.stringify(body) : undefined,
    signal: AbortSignal.timeout(20000),
  });
  const payload = await response.json().catch(() => ({}));
  if (!response.ok) {
    const detail = payload?.error?.message ?? response.statusText;
    throw new Error(`Graph API ${response.status}: ${detail}`);
  }
  return payload;
}

export async function sendText(to, text, { previewUrl = true } = {}) {
  return graphRequest(`${config.meta.phoneNumberId}/messages`, {
    body: {
      messaging_product: 'whatsapp',
      recipient_type: 'individual',
      to,
      type: 'text',
      text: { preview_url: previewUrl, body: text },
    },
  });
}

export async function markRead(messageId) {
  if (!config.meta.markAsRead || !messageId) return null;
  try {
    return await graphRequest(`${config.meta.phoneNumberId}/messages`, {
      body: { messaging_product: 'whatsapp', status: 'read', message_id: messageId },
    });
  } catch (error) {
    logger.debug('تعذر تعليم الرسالة كمقروءة', { error: error.message });
    return null;
  }
}

/** Two-step download: resolve the media id to a URL, then fetch the bytes. */
export async function downloadMedia(mediaId) {
  const meta = await graphRequest(mediaId, { method: 'GET' });
  if (!meta?.url) throw new Error('لم يُرجع Graph API رابط الوسائط');
  const response = await fetch(meta.url, {
    headers: { authorization: `Bearer ${config.meta.token}` },
    signal: AbortSignal.timeout(30000),
  });
  if (!response.ok) throw new Error(`تعذر تنزيل الوسائط: ${response.status}`);
  const buffer = Buffer.from(await response.arrayBuffer());
  logger.debug('تم تنزيل وسائط', { id: redact(mediaId), bytes: buffer.length });
  return { data: buffer.toString('base64'), mediaType: meta.mime_type ?? 'image/jpeg' };
}

export const metaProvider = {
  name: 'meta',
  sendText,
  markRead,
  downloadMedia,
  verifySignature,
  parseWebhook,
};

File: app/whatsapp/meta.py
"""مزوّد Meta WhatsApp Cloud API (الرسمي)."""

from __future__ import annotations

import logging

import httpx

from ..config import normalize_phone, settings
from .base import InboundMessage, WhatsAppProvider, split_message, to_whatsapp_markdown

log = logging.getLogger(__name__)

GRAPH_VERSION = "v21.0"


class MetaProvider(WhatsAppProvider):
    name = "meta"

    def __init__(self, token: str | None = None, phone_number_id: str | None = None) -> None:
        self.token = token or settings.meta_access_token
        self.phone_number_id = phone_number_id or settings.meta_phone_number_id
        self.base_url = f"https://graph.facebook.com/{GRAPH_VERSION}/{self.phone_number_id}/messages"

    # ---------------------------------------------------------------- الإرسال

    def _post(self, body: dict) -> str | None:
        if not self.token or not self.phone_number_id:
            log.error("إعدادات Meta ناقصة (META_ACCESS_TOKEN / META_PHONE_NUMBER_ID)")
            return None
        headers = {"Authorization": f"Bearer {self.token}", "Content-Type": "application/json"}
        try:
            with httpx.Client(timeout=30.0) as client:
                response = client.post(self.base_url, json=body, headers=headers)
            if response.status_code >= 400:
                log.error("رفض من واتساب (%s): %s", response.status_code, response.text[:500])
                return None
            data = response.json()
            return (data.get("messages") or [{}])[0].get("id")
        except httpx.HTTPError as exc:
            log.error("فشل الاتصال بواتساب: %s", exc)
            return None

    def send_text(self, phone: str, text: str) -> list[str]:
        phone = normalize_phone(phone)
        ids: list[str] = []
        for chunk in split_message(to_whatsapp_markdown(text)):
            message_id = self._post({
                "messaging_product": "whatsapp",
                "recipient_type": "individual",
                "to": phone,
                "type": "text",
                "text": {"preview_url": False, "body": chunk},
            })
            if message_id:
                ids.append(message_id)
        return ids

    def send_template(self, phone: str, template: str, params: list[str]) -> str | None:
        return self._post({
            "messaging_product": "whatsapp",
            "to": normalize_phone(phone),
            "type": "template",
            "template": {
                "name": template,
                "language": {"code": settings.meta_template_lang},
                "components": [{
                    "type": "body",
                    "parameters": [{"type": "text", "text": p} for p in params],
                }] if params else [],
            },
        })

    # ---------------------------------------------------------------- الاستقبال

    def verify_webhook(self, mode: str, token: str, challenge: str) -> str | None:
        """تحقق Meta من الـ webhook عند الربط الأولي."""
        if mode == "subscribe" and token == settings.meta_verify_token:
            return challenge
        return None

    def parse_webhook(self, payload: dict) -> list[InboundMessage]:
        messages: list[InboundMessage] = []
        for entry in payload.get("entry", []):
            for change in entry.get("changes", []):
                value = change.get("value", {})
                # خريطة الأرقام إلى الأسماء من حقل contacts
                names = {
                    c.get("wa_id"): (c.get("profile") or {}).get("name")
                    for c in value.get("contacts", [])
                }
                for message in value.get("messages", []):
                    phone = normalize_phone(message.get("from", ""))
                    kind = message.get("type", "unsupported")
                    if kind == "text":
                        text = (message.get("text") or {}).get("body", "")
                    elif kind == "button":
                        text = (message.get("button") or {}).get("text", "")
                    elif kind == "interactive":
                        interactive = message.get("interactive") or {}
                        text = (
                            (interactive.get("button_reply") or {}).get("title")
                            or (interactive.get("list_reply") or {}).get("title")
                            or ""
                        )
                    else:
                        text = ""
                    messages.append(InboundMessage(
                        phone=phone,
                        text=text,
                        name=names.get(message.get("from")),
                        message_id=message.get("id"),
                        kind=kind if kind in ("text", "audio", "image") else "unsupported",
                    ))
        return messages
